payments/todopago/__init__.py

```diff
@@ -0,0 +1,147 @@
+                    item.quantity,
+                    round(item.price * Decimal(item.quantity)),
+                )
+                for item in payment.get_purchased_items()
+            ],
+        )
+
+        if authorization.status_code != -1:
+            raise PaymentError(
+                message=f"Failed to authorize TodoPago operation. [{authorization.status_message}]",
+                code=authorization.status_code,
+                gateway_message=authorization.status_message,
+            )
+
+        payment.attrs.request_key = authorization.request_key
+        payment.attrs.public_request_key = authorization.public_request_key
+        payment.attrs.form_url = authorization.form_url
+
+        payment.transaction_id = transaction_id
+        payment.save()
+
+        return authorization
+
+    def fetch_operation_status(self, payment: BasePayment):
+        request_key = payment.attrs.request_key
+        answer_key = payment.attrs.answer_key
+
+        status = self.client.get_operation_status(request_key, answer_key)
+
+        if status.status_code != -1:
+            message = "TodoPago didn't approve the payment."
+            payment.change_status(PaymentStatus.ERROR, message)
+            raise (PaymentError(message))
+
+        upstream_status = status.status_message
+        payment.change_status(STATUS_MAP[upstream_status])
+
+    def get_action(self, payment: BasePayment):
+        # This is the form-action. But we don't use a form.
+        raise NotImplementedError()
+
+    def process_callback(self, payment: BasePayment, request: HttpRequest):
+        """Process TodoPago's callback.
+
+        After a payment is made, TodoPago sends a GET request with the
+        answer key. We use this to get the status of the payment.
+
+        If there's no answer key then we assume that the payment failed.
+        """
+        answer_key = request.GET.get("Answer", None)
+        if not answer_key:
+            payment.change_status(PaymentStatus.ERROR)
+            return redirect(payment.get_failure_url())
+
+        payment.attrs.answer_key = answer_key
+        payment.save()
+
+        self.fetch_operation_status(payment)
+
+        return redirect(payment.get_success_url())
+
+    def process_data(self, payment: BasePayment, request: HttpRequest):
+        """Handle a request received after a payment.
+
+        If it's a GET request, then it's the user being redirected after
+        completing a payment (it may have failed or have been successfull).
+
+        TodoPago doesn't send POST requests, at least for now.
+        If the payment is refunded from the TodoPago's platform there's no way
+        we notice that.
+        """
+        if request.method == "GET":
+            return self.process_callback(payment, request)
+
+    def get_form(self, payment: BasePayment, data=None):
+        if not hasattr(payment.attrs, "form_url"):
+            _ = self.authorize_operation(payment)
```

Comment:
Why do you assign to `_` here?

Reply:
I misunderstood the purpose of `_`.  Here is [pep640](https://peps.python.org/pep-0640/) specification for unused variable syntax, this would make sense for something like `a, _ = some_func()` where the second return value is not being used.
I'll remove them.

Reply:
Using `_` is for cases when the assignment is unaviodable, like:

```py
for _ in range(0, 10):
   print("waiting...")
```

Or

```py
x, _ = func_that_returns_two_values()
print(x)
```

Reply:
Thanks for the explanation, I'll keep that in mind.